addons/l10n_in/models/product_template.py

```diff
@@ -1,11 +1,20 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import models, fields
+from odoo import api, models, fields, _
+from odoo.exceptions import ValidationError
 
 
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
     l10n_in_hsn_code = fields.Char(string="HSN/SAC Code", help="Harmonized System Nomenclature/Services Accounting Code")
     l10n_in_hsn_description = fields.Char(string="HSN/SAC Description", help="HSN/SAC description is required if HSN/SAC code is not provided.")
+
+    @api.constrains('l10n_in_hsn_code')
+    def _check_hsn_code_validation(self):
+        for record in self:
+            company = self.env.company
+            if record.l10n_in_hsn_code and company.l10n_in_hsn_validation and str(len(record.l10n_in_hsn_code)) < company.l10n_in_hsn_code_digit:
```

Comment:
```suggestion
            if company.country_id.code == 'IN' and record.l10n_in_hsn_code and company.l10n_in_hsn_validation and len(record.l10n_in_hsn_code) < int(company.l10n_in_hsn_code_digit or '8'):
```

you are comparing string for length check !!!